File: files/2026-06-14/CSEC 14062026.py
import pandas as pd
import calendar

path = "2026-06-14/Challenge 140.xlsx"
input = pd.read_excel(path, usecols="B:C", nrows=7, skiprows=1)
test = pd.read_excel(path, usecols="E:F", nrows=2, skiprows=1).rename(
    columns=lambda c: c.replace(".1", "")
)

month_abb = list(calendar.month_abbr[1:])
month_map = {m: i + 1 for i, m in enumerate(month_abb)}
input_m = input.assign(Month=input["Month"].map(month_map))
result = (
    input_m[["Customer", "Month"]]
    .drop_duplicates()
    .assign(Month=lambda df: df["Month"] + 1)
    .merge(
        input_m[["Customer", "Month"]],
        on=["Customer", "Month"],
        how="left",
        indicator=True,
    )
    .loc[lambda df: df["_merge"] == "left_only", ["Customer", "Month"]]
)
max_month = input["Month"].map(month_map).max(skipna=True)
result = (
    result.loc[result["Month"] <= max_month]
    .assign(
        Month=lambda df: df["Month"].apply(
            lambda m: month_abb[int(m) - 1] if pd.notna(m) else pd.NA
        ),
        **{"Customer Lost": lambda df: df["Customer"]},
    )[["Month", "Customer Lost"]]
    .reset_index(drop=True)
)
print(result.equals(test))
# True

File: files/2026-06-14/CSEC 14062026.R
library(tidyverse)
library(readxl)

path <- "2026-06-14/Challenge 140.xlsx"
input <- read_excel(path, range = "B2:C8")
test <- read_excel(path, range = "E2:F4")

result <- input %>%
  mutate(Month = match(Month, month.abb)) %>%
  distinct(Customer, Month) %>%
  transmute(Customer, Month = Month + 1) %>%
  anti_join(
    input %>% mutate(Month = match(Month, month.abb)),
    by = c("Customer", "Month")
  ) %>%
  filter(Month <= max(match(input$Month, month.abb), na.rm = TRUE)) %>%
  transmute(Month = month.abb[Month], `Customer Lost` = Customer)

all.equal(result, test)
# [1] TRUE
